BM/tools/pcie/pcie_check.c

```diff
@@ -0,0 +1,1225 @@
+	printf("x    Only check CXL related registers:x 4 16 1e98\n");
+	printf("X    Only check CXL related registers value included:X 4 16 8\n");
+	printf("v    Verify PCIe register:v 23 4 16 1e98\n");
+	printf("V    Verify PCIe register was included:V 23 4 16 8\n");
+	printf("w    Write PCIe register if writeable:w 12 8 16 11\n");
+	printf("bus  Specific bus number(HEX)\n");
+	printf("dev  Specific device number(HEX)\n");
+	printf("func Specific function number(HEX-optional)\n");
+	printf("Write specific pcie 6b:00.0 reg sample:w 23 20 32 0002 6b 00 0\n");
+	exit(2);
+}
+
+unsigned long find_base_from_dmesg(void)
+{
+	FILE *fp;
+	unsigned long base_addr = 0;
+	char result[256];
+	const char *cmd = "dmesg | grep 'MMIO range' | grep 'Remove' | head -n 1";
+	char *start;
+
+	fp = popen(cmd, "r");
+	if (!fp) {
+		printf("Failed to run dmesg command by popen\n");
+		return base_addr;
+	}
+
+	if (fgets(result, sizeof(result), fp)) {
+		start = strstr(result, "[0x");
+		if (start) {
+			if (sscanf(start, "[0x%lx-", &base_addr) == 1)
+				printf("MMIO BASE from dmesg:0x%lx\n", base_addr);
+			else
+				printf("No MMIO BASE from dmesg:0x%lx\n", base_addr);
+		} else {
+			printf("No MMIO range in dmesg:0x%lx\n", base_addr);
+		}
+	} else {
+		printf("No useful MMIO info in dmesg.\n");
+		if (pclose(fp) == -1) {
+			perror("pclose failed");
+			return base_addr;
+		}
+		return base_addr;
+	}
+
+	if (pclose(fp) == -1) {
+		perror("pclose failed");
+		return base_addr;
+	}
+
+	return base_addr;
+}
+
+unsigned long find_base_from_mcfg(void)
+{
+	FILE *fp;
+	unsigned long base_addr = 0;
+	// 16 chars actually and 64 is enough
+	char result[64];
+	const char *cmd = "b=$(ls /sys/firmware/acpi/tables/MCFG* | head -n 1);"
+			"a=$(hexdump $b | grep 000030 | awk '{print $3 $2}');"
+			"a+=$(hexdump $b | grep 000020 | awk '{print $9 $8}');"
+			"echo $a";
+	fp = popen(cmd, "r");
+	if (!fp) {
+		printf("Failed to run cmd command by popen\n");
+		return base_addr;
+	}
+
+	if (fgets(result, sizeof(result), fp)) {
+		if (sscanf(result, "%lx", &base_addr) == 1)
+			printf("MMIO BASE from sysfs mcfg:0x%lx\n", base_addr);
+		else
+			printf("No MMIO BASE from sysfs mcfg:0x%lx\n", base_addr);
+	} else {
+		printf("Failed to read mcfg sysfs by fgets\n");
+		if (pclose(fp) == -1) {
+			perror("pclose failed");
+			return base_addr;
+		}
+		return base_addr;
+	}
+
+	if (pclose(fp) == -1) {
+		perror("pclose failed");
+		return base_addr;
+	}
+
+	return base_addr;
+}
+
+int find_bar(void)
+{
+#ifdef __x86_64__
+	FILE *maps;
+	unsigned long *start = malloc(sizeof(unsigned long) * 5);
+	char line[MAPS_LINE_LEN], base_end[MAPS_LINE_LEN];
+	char *mmio_bar = "MMCONFIG";
+
+	//printf("Try to open /proc/iomem\n");
+	maps = fopen("/proc/iomem", "r");
+	if (!maps) {
+		printf("[WARN]\tCould not open /proc/iomem\n");
+		exit(1);
+	}
+
+	while (fgets(line, MAPS_LINE_LEN, maps)) {
+		if (!strstr(line, mmio_bar))
+			continue;
+
+		if (sscanf(line, "%p-%s", &start, base_end) != 2)
```

Comment:
Hi @ysun,

Here must be &start, to fetch the pointer address, and then start could fetch the pointer value correctly.
Otherwise, it will get the wrong result.
Should not change.

Thanks!